packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTree.tsx

```diff
@@ -119,6 +133,100 @@ export function ModelsTree(props: ModelsTreeProps) {
   );
 }
 
+interface ModelsTreeRendererProps extends TreeRendererBaseProps {
+  nodeRendererProps: ModelsTreeNodeRendererProps;
+}
+
+function createModelsTreeRenderer(
+  { nodeRendererProps, ...restProps }: ModelsTreeRendererProps,
+  nodeLoader: AbstractTreeNodeLoaderWithProvider<IPresentationTreeDataProvider>,
+) {
+  return function ModelsTreeRenderer(treeProps: TreeRendererProps) {
+    const { onContextMenu, renderContextMenu } = useContextMenu({ contextMenuItems: restProps.contextMenuItems });
+    const { onClearFilterClick, onFilterClick, filterDialog, containerRef } = useFilterablePresentationTree<HTMLDivElement>({ nodeLoader });
+    const isEnlarged = restProps.density === "enlarged";
+    const className = classNames("tree-widget-tree-nodes-list", { ["enlarge"]: isEnlarged });
+
+    if (isEnlarged) {
+      treeProps.nodeHeight = () => 43;
+    }
+
+    return (
+      <div className={className} ref={containerRef}>
+        <TreeRenderer
+          {...restProps}
+          {...treeProps}
+          nodeRenderer={createModelsTreeNodeRenderer(
+            { ...nodeRendererProps, isEnlarged },
+            {
+              onClearFilterClick,
+              onFilterClick,
+              onContextMenu,
+            },
+          )}
+        />
+        {renderContextMenu()}
+        {filterDialog}
+      </div>
+    );
+  };
+}
+
+interface ModelsTreeNodeActions {
+  onClearFilterClick: (nodeId: string) => void;
+
+  onFilterClick: (nodeId: string) => void;
+
+  onContextMenu: (e: React.MouseEvent, node: TreeModelNode) => void;
+}
+
+interface ModelsTreeNodeRendererProps {
+  iconsEnabled: boolean;
+
+  descriptionEnabled: boolean;
+
+  levelOffset: number;
+
+  disableRootNodeCollapse: boolean;
+
+  isEnlarged?: boolean;
+}
+
+/**
+ * Creates node renderer which renders node with eye checkbox.
+ * @public
+ */
+function createModelsTreeNodeRenderer(
```

Comment:
I think this is an opportunity to flatten these `create*` functions. I would suggest to remove it and leave only `ModelsTreeNodeRenderer` something like:

```tsx
interface ModelsTreeNodeRendererProps extends PresentationTreeNodeRenderer {
    iconsEnabled: boolean;
    descriptionEnabled: boolean
    levelOffset: number;
    // ...
}

function ModelsTreeNodeRenderer({iconsEnabled, descriptionEnabled, levelOffset, ...props}: ModelsTreeNodeRendererProps) {
    // ...
}
```